### ISSUE TITLE
ld seems to have an old PE stub

### ISSUE THREAD

**Issue by stsp** [jwt27/djgpp-cvs#6]
Hi.

I wasn't yet successful in creating
PE executable with raw binutils
(rather than with djgpp), but, unless
I am mistaken, ld still embeds the
old stub (the one w/o my patches).
Only stubify tool has the new stub,
so when using djgpp, the stub is good.
But I believe creating PE with raw
binutils will yield the old stub.
I looked into an ld binary and I can
see the old stub there because the
error msg that I added, is not there.

How does ld embed the PE stub?
Is it hard-coded somewhere in the
sources? I mean, the one doesn't
build djgpp libc when building the
raw binutils. But the stub is there.

**Comment by stsp** [comment]
OK, it you think updating the stub
is not needed, then lets close.
I am just pondering around these
tools and am trying to find the minimal
set need to create the dos coff executable
from an asm file, without involving
djgpp gcc.

**Comment by stsp** [comment]
Btw how feasible is to convert an elf object
to coff object? Current objcopy doesn't seem
to be able to do that...

**Comment by stsp** [comment]
> Not sure what you mean though - PE executables are for windows (or HX) and have no djgpp stub at all.

Indeed but the object files produced
with djgpp are very similar to the object
files produced with `ld -mi386pe` or with
objcopy from elf to PE.
I can even use `nm` from casual
binutils with djgpp-created object files.
Can I assume that only executables
would be different, but the object files
can be created from asm with raw
binutils?

**Comment by stsp** [comment]
[djgpp.tar.gz](https://github.com/jwt27/djgpp-cvs/files/10046634/djgpp.tar.gz)

Well, no, it _is_ a bug.
Attached the test-case.
Just run `make` and you'll get a
perfectly working binary but with
an old stub.

**Comment by jwt27** [comment]
> Can I assume that only executables
> would be different, but the object files
> can be created from asm with raw
> binutils?

No, I don't think that's safe.  I don't have an exact list of all the differences, but at least the section names and flags are incompatible.  Compare the linker script templates for [PE](https://github.com/bminor/binutils-gdb/blob/master/ld/scripttempl/pe.sc) and [go32](https://github.com/bminor/binutils-gdb/blob/master/ld/scripttempl/i386go32.sc).
It may happen to work for a basic hello world with only `.text`/`.data`/`.bss`, but with anything more complicated you'll likely run into trouble.


> Well, no, it _is_ a bug.
> Attached is the test-case.
> Just run `make` and you'll get a
> perfectly working binary but with
> an old stub.

You'll get the new stub if you use:
```
DOS_LD ?= i386-pc-msdosdjgpp-gcc`
```
That way you also don't need to link with `crt0.o` and `libc.a` manually.

You can even assemble `prnt.S` via gcc:
```
%.o : %.S
	$(DOS_CC) -c $< -o $@
```
It will do the right thing based on the file extension.

**Comment by stsp** [comment]
The idea was to avoid djgpp-gcc and
build basic asm stuff with just djgpp-binutils
that are in both debian and ubuntu.
